File: Software Development Module/Revision/Exam2016/PasswordCheckerApp.java
/*
PasswordCheckerApp.java
Billy Cussen
13/08/2020
*/

import java.util.Scanner;

//Write a Java App takes in as a parameter a one-dimensional array of Strings which contains passwords
//The method should search (i.e. traverse) the array of passwords and display all the passwords which are weak.

//Password rules:
//A password must contain at least 10 characters.
//A password must consist of letters, digits & Characters.
//A password should contain at least two Characters.
//A password should contain at least one uppercase letter

import java.util.InputMismatchException;

public class PasswordCheckerApp{

    public static void main(String [] args){

        //Variables
        int passwordAmount = 0;
        String[] passwordsArray;
        String output;
        String specialCharacters = "/*!@#$%^&'*()\"{}_[]|\\?/<>,.";
        boolean inputCheck;

        //Objects
        PasswordChecker myPasswordChecker;
        Scanner keyboard;

        //Input
        do{
            keyboard = new Scanner(System.in);
            inputCheck = true;

            try{    
                System.out.print("Please enter the number of Passwords to be Checked: ");
                passwordAmount = keyboard.nextInt();

                if(passwordAmount <= 0 || passwordAmount > 10){
                    throw new InputMismatchException();
                }

            }catch(NumberFormatException | InputMismatchException e){
                System.out.println("Please enter a Valid number (e.g. 3)\nPlease remember you can only try a maximum of 10 Passwords per Check");
                inputCheck = false;
            }finally{
                keyboard = null;
            }

        }while(!inputCheck);

        passwordsArray = new String[passwordAmount];

        for(int i = 0; i < passwordsArray.length; i++){

            
            do{
                keyboard = new Scanner(System.in);
                inputCheck = true;

                try{    
                    System.out.print("Please enter Password "+(i+1)+": ");
                    passwordsArray[i] = keyboard.nextLine();

                    if(passwordsArray[i].length()<10){
                        throw new InvalidPasswordException();
                    }

                    boolean digitCheck = false;
                    boolean letterCheck = false;
                    boolean characterCheck = false;

                    for(int j = 0; j < passwordsArray[i].length(); j++){

                        if(Character.isDigit(passwordsArray[i].charAt(j))){
                            digitCheck = true;
                        } else if(Character.isLetter(passwordsArray[i].charAt(j))){
                            letterCheck = true;
                        }

                        for(int k = 0; k < specialCharacters.length();k++){

                            if(passwordsArray[i].charAt(j) == specialCharacters.charAt(k)){
                                characterCheck = true;
                                break;
                            }

                        }

                        if(digitCheck&&letterCheck&&characterCheck){
                            break;
                        }
                    }

                    if(!(digitCheck&&letterCheck&&characterCheck)){
                        throw new InvalidPasswordException("Password "+(i+1)+" requires at least 1 Letter, Number & Character!");
                    }

                }catch(InvalidPasswordException f){
                    System.out.println(f.getMessage());
                    inputCheck = false;
                }finally{
                    keyboard = null;
                }

            }while(!inputCheck);

        }

        //Process
        myPasswordChecker = new PasswordChecker(passwordsArray, specialCharacters);
        output = myPasswordChecker.getWeakPasswords();
        myPasswordChecker = null;
        //Output
        System.out.println(output);

    }
}

File: Python Course/Software Development Module - in Python/Revision/Exam2016/PasswordChecker/PasswordCheckerApp.py
"""
PasswordCheckerApp.py
Billy Cussen
31/10/2020
"""

#Write a Java(Python here) App takes in as a parameter a one-dimensional array of Strings which contains passwords
#The method should search (i.e. traverse) the array of passwords and display all the passwords which are weak.

#Password rules:
#A password must contain at least 10 characters.
#A password must consist of letters, digits & Characters.
#A password should contain at least two Characters.
#A password should contain at least one uppercase letter

#Object
import gc
import PasswordChecker
from PasswordChecker import myPasswordChecker

#Variables
repeat = "y"
passwordsList = []

while repeat == "y":

    #Input - Ticket Type
    inputCheck = False
    while inputCheck == False:
        try:
            numberOfPasswords = input("Please enter the number of Passwords to be checked: ")
            numberOfPasswords = int(numberOfPasswords)
            if numberOfPasswords <= 0 or numberOfPasswords > 10:
                raise Exception
        except Exception:
            print("Please enter a Valid number (e.g. 3)\nPlease remember you can only try a maximum of 10 Passwords per Check")
        else:
            inputCheck = True

    for i in range(numberOfPasswords):
        inputCheck = True
        while inputCheck:
            try:
                password = input(str("Please enter Password "+str(i+1)+": "))
                if len(password) < 10:
                    raise Exception
            except Exception:
                print("Password"+str(i+1)+" needs at least 10 characters!")
            else:
                passwordsList.append(password)
                inputCheck = False


    #Process & Output
    myPasswordCheckerApp = myPasswordChecker(passwordsList)
    try:
        print(myPasswordCheckerApp.getPasswords())
    except Exception as e:
        print(e)
    else:
        myPasswordCheckerApp = None
        passwordsList = []
        gc.collect

    #Input - Repeat?
    repeat = ""
    while repeat != "y" and repeat != "n":
        repeat = input(str("Please enter \"y\" to retry, enter \"n\" to exit: "))
        repeat = repeat.lower()
        if repeat != "y" and repeat != "n":
            print("You need to enter \"y\" or \"n\"")

print("Thanks for using my Password Checker App!")